libres/tests/analysis/test_update.cpp

```diff
@@ -0,0 +1,341 @@
+b_f = rng.normal(loc=b_t + 0.5 * b_f_std, scale=b_f_std)
+
+def model(x, a, b):
+    return a * x + b
+
+obs_std = 0.5
+
+# Measurement matrix M is the linear model implemented as a matrix.
+M = np.ones(shape=(nparam, nparam))
+M[0, 0] = 0
+
+psi_f = np.array([a_f, b_f])
+C_f_psi = np.identity(nparam)
+C_f_psi[0, 0] = a_f_std ** 2
+C_f_psi[1, 1] = b_f_std ** 2
+
+C_eps = np.identity(nparam) * obs_std ** 2
+
+niter = 1
+
+for n in range(niter):
+    K = C_f_psi @ M.T @ np.linalg.inv(M @ C_f_psi @ M.T + C_eps)
+    d = np.array(
+        [
+            model(0, a_t, b_t) + rng.normal(loc=0, scale=obs_std),
+            model(1, a_t, b_t) + rng.normal(loc=0, scale=obs_std),
+        ]
+    )
+    psi_a = psi_f + K @ (d - M.dot(psi_f))
+    psi_a = psi_f + K @ (d - M.dot(psi_f))
+    C_a_psi = (np.identity(nparam) - K @ M) @ C_f_psi
+    psi_f = psi_a
+    C_f_psi = C_a_psi
+
+a_a = psi_a[0]
+b_a = psi_a[1]
+a_a_std = C_a_psi[0, 0]
+b_a_std = C_a_psi[1, 1]
+*/
+
+// Global variables used to test behaviour of series of estimates
+std::vector<double> a_diff_ml_posterior;
+std::vector<double> b_diff_ml_posterior;
+std::vector<double> covariance_determinant;
+SCENARIO("Running analysis update without row scaling on linear model",
+         "[analysis]") {
+
+    GIVEN("Different setups for analysis update") {
+        analysis_module_type *enkf_module = analysis_module_alloc("STD_ENKF");
+        // Default truncation of 0.98 of eigenvalue spectrum is such that this
+        // test will fail unless we explicitly set it to 1.0.
+        analysis_module_set_var(enkf_module, ENKF_TRUNCATION_KEY_, "1.0");
+
+        int ens_size = GENERATE(10, 100, 1000);
+        int obs_size = 45;
+        double obs_std = GENERATE(100.0, 1.0, 0.5, 0.1, 0.01, 0.00001);
+
+        std::vector<model> ens;
+
+        bool_vector_type *ens_mask = bool_vector_alloc(ens_size, true);
+        obs_data_type *obs_data = obs_data_alloc(1.0);
+        meas_data_type *meas_data = meas_data_alloc(ens_mask);
+        rng_type *rng = rng_alloc(MZRAN, INIT_DEFAULT);
+
+        model true_model{a_true, b_true};
+
+        for (int iens = 0; iens < ens_size; iens++)
+            ens.push_back(model(rng));
+
+        const char *obs_key = "OBS1";
+        meas_block_type *mb =
+            meas_data_add_block(meas_data, obs_key, 1, obs_size);
+        obs_block_type *ob =
+            obs_data_add_block(obs_data, obs_key, obs_size, nullptr, false);
+
+        std::vector<double> xarg(obs_size);
+
+        // Useful example of how to add measurements to only two points,
+        // in this case x=0 and x=1.
+        // std::fill(xarg.begin(), xarg.begin() + obs_size / 2, 0);
+        // std::fill(xarg.begin() + (obs_size - obs_size / 2), xarg.end(), 1);
+
+        for (int i = 0; i < obs_size; i++) {
+            xarg[i] = i;
+        }
+
+        for (int iens = 0; iens < ens_size; iens++) {
+            const auto &m = ens[iens];
+            // This is equivalent to M*psi_f in the book
```

Comment:
This is the scientist in me talking - but which book? I know you are refering to the one referenced at the top of the file, but maybe add a reference?

Reply:
I've added a reference, thanks.